h2/src/test/org/h2/test/db/TestSynonymForTable.java

```diff
@@ -0,0 +1,321 @@
+        testCreateForUnknownTable();
+        testMetaData();
+        testCreateOrReplace();
+        testCreateOrReplaceExistingTable();
+        testSynonymInDifferentSchema();
+        testReopenDatabase();
+        testDropSynonym();
+        testDropTable();
+        testDropSchema();
+    }
+
+    private void testUpdate() throws SQLException {
+        Connection conn = getConnection("synonym");
+        createTableWithSynonym(conn);
+        insertIntoSynonym(conn, 25);
+
+        Statement stmnt = conn.createStatement();
+        assertEquals(1, stmnt.executeUpdate("UPDATE testsynonym set id = 30 WHERE id = 25"));
+
+        assertSynonymContains(conn, 30);
+
+        conn.close();
+    }
+
+    private void testDropSchema() throws SQLException {
+        Connection conn = getConnection("synonym");
+        Statement stat = conn.createStatement();
+
+        stat.execute("CREATE SCHEMA IF NOT EXISTS s1");
+        stat.execute("CREATE TABLE IF NOT EXISTS s1.backingtable(id INT PRIMARY KEY)");
+        stat.execute("CREATE OR REPLACE SYNONYM testsynonym FOR s1.backingtable");
+        stat.execute("DROP SCHEMA s1");
+
+        assertThrows(JdbcSQLException.class, stat).execute("SELECT id FROM testsynonym");
+        conn.close();
+    }
+
+    private void testDropTable() throws SQLException  {
+        Connection conn = getConnection("synonym");
+        createTableWithSynonym(conn);
+        Statement stat = conn.createStatement();
+        stat.execute("DROP TABLE backingtable");
+
+        // Backing table does not exist anymore.
+        assertThrows(JdbcSQLException.class, stat).execute("SELECT id FROM testsynonym");
+
+        // Meta data should show INVALID
+        ResultSet synonyms = conn.createStatement().executeQuery("SELECT * FROM INFORMATION_SCHEMA.SYNONYMS WHERE SYNONYM_NAME='TESTSYNONYM'");
+        assertTrue(synonyms.next());
+        assertEquals("TESTSYNONYM", synonyms.getString("SYNONYM_NAME"));
+        assertEquals("INVALID", synonyms.getString("STATUS"));
+        conn.close();
+
+        // Reopending should work with invalid synonym
+        Connection conn2 = getConnection("synonym");
+        assertThrows(JdbcSQLException.class, stat).execute("SELECT id FROM testsynonym");
+        conn2.close();
+    }
+
+    private void testDropSynonym() throws SQLException {
+        Connection conn = getConnection("synonym");
+        createTableWithSynonym(conn);
+        Statement stat = conn.createStatement();
+
+        stat.execute("DROP SYNONYM testsynonym");
+
+        // Synonym does not exist anymore.
+        assertThrows(JdbcSQLException.class, stat).execute("SELECT id FROM testsynonym");
+
+        // Dropping with "if exists" should succeed even if the synonym does not exist anymore.
+        stat.execute("DROP SYNONYM IF EXISTS testsynonym");
+
+        // Without "if exists" the command should fail.
+        assertThrows(JdbcSQLException.class, stat).execute("DROP SYNONYM testsynonym");
```

Comment:
Well, otherwise the "IF EXISTS" does not make a lot of sense. I basically wanted this to have the same semantics as "DropTable" where dropping a non-existing table also causes an error if "IF EXISTS" is not specified.